homeassistant/components/mqtt/client.py

```diff
@@ -895,6 +895,12 @@ def _async_mqtt_on_connect(
         import paho.mqtt.client as mqtt
 
         if result_code != mqtt.CONNACK_ACCEPTED:
+            if result_code in (
+                mqtt.CONNACK_REFUSED_BAD_USERNAME_PASSWORD,
+                mqtt.CONNACK_REFUSED_NOT_AUTHORIZED,
+            ):
+                self.hass.async_create_task(self.async_disconnect())
+                self.config_entry.async_start_reauth(self.hass)
             _LOGGER.error(
```

Comment:
This is probably an existing problem, but I think we need to call `self._async_connection_result(False)` before the return

Reply:
https://github.com/bdraco/home-assistant/commit/fa7430586ebfe339e5d8f0883c78b5548ac14b0a